compiler/rustc_codegen_ssa/src/sir.rs

```diff
@@ -411,6 +411,18 @@ impl SirFuncCx<'tcx> {
                 self.lower_binop(bx, bb, dest_ty, *op, opnd1, opnd2, true)
             }
             mir::Rvalue::Cast(mir::CastKind::Misc, op, ty) => self.lower_cast_misc(bx, bb, op, ty),
+            mir::Rvalue::Len(p) => {
+                // Yorick's register allocator will always put a fat pointer on the stack, so
```

Comment:
What does "always put a fat pointer on the stack" mean? A struct that's something like `struct S {x: &dyn T}` could be on the heap.

Reply:
Currently the JIT doesn't understand the heap (will crash trying to compile a `Box`). It only knows that since the value is 128-bit, it can't fit in a register. You are right though, we should future proof this comment. How about:

> For now, Yorick's register allocator will always put fat pointers in memory, ...

One optimisation that could come later is to relax this constraint, and allow 128-bit values to span two-registers. But now isn't the time, I don't think.

Reply:
"For now, Yorick's register allocator will always put fat pointers in memory" is very easily misunderstood. Maybe this should be more of a "FIXME: currently this relies on the fact that yk's register allocator always stores fat pointers on the stack". An `assert`, if possible, would be an even better alternative!

Reply:
The thing is that it's not clear if it's something we'd want to "fix" without benchmarking.

What would you be asserting? Register allocation decisions happen much later at trace compile time.

Reply:
So how about:

> this relies on the fact that yk's register allocator never stores fat pointers in registers.

Reply:
> The thing is that it's not clear if it's something we'd want to "fix" without benchmarking.

I think you can safely assume that, if used in a tight loop, anything (including, but not limited to, fat pointers) will benefit from being in registers.

> What would you be asserting?

Can't we assert that it actually *has* been stored on the stack and not elsewhere?

Reply:
Afraid not, because no register allocation decisions are made at SIR lowering time. Registers are allocated during trace code-generation.

Reply:
Right I see. Obvious other question, then: can we calculate its length another / better way?

Reply:
As I understand, the reason that Rust does it this way is that the length of a slice can't be known statically. Otherwise there'd be no need for a fat pointer and a constant could be inserted (as it appears they do for arrays, whose length, unlike slices, is encoded in its type).

Reply:
The size of a slice is constant (two machine words). The second element tells you how many items it points to, which is rather different. I'm not sure which we're talking about now.

Reply:
You lost me.

I thought you were asking if we could compute the second element of a fat pointer statically. I don't think you can.

Reply:
Did you mean, can we make it so that ykrustc knows what could live in a register at runtime?

That could be done if we share `can_live_in_register()` somehow.

Reply:
> I thought you were asking if we could compute the second element of a fat pointer statically. I don't think you can.

There are two sized things here: how many bytes does the *slice* occupy? how many items does the slice point to? The latter is dynamic and thus unknowable (in general) at compile-time. The former is static and knowable at compile-time (if it wasn't, we wouldn't be able to compile anything). I'm trying to work out which of these two things we're getting the size of here.

Reply:
`Rvalue::len`, the MIR bytecode, accesses the length of the collection pointed to (be it array or slice, although this change addresses the slice case).

We are lowering such a bytecode to a SIR bytecode that accesses the fat pointer + 8.

I'm starting to wonder if my original comment is a bit misguided. Even if the code-gen could span a 128-bit value over 2 registers, we could still make the `+8` work by considering two registers as a "paired" (i.e. we consider two registers (e.g.) `rax` and `rbx` as contiguous storage, and reading past `rax` leads into `rbx`).

What do you think? Does that make sense? Assuming it does, I suggest I remove the comment altogether, as in hindsight it seems to do more harm than good.

Reply:
I don't think we would necessarily put the two values in consecutive registers (and does it even make sense to view registers as storage?!).

It feels like the right comment might be "FIXME: this won't work correctly if the second element of the slice is not stored immediately after the first"?

Reply:
> I don't think we would necessarily put the two values in consecutive registers.

I'm not saying the registers would need to be consecutive, just that the allocator could in theory issue a pair of registers (any two) and consider them a 128-bit space. Rust does exactly this over the ABI boundary for 128-bit values. A fat pointer is split between two registers if possible.

> does it even make sense to view registers as storage?

I'm not sure what the controversy is there :P When I `mov` a value into a register, it can be thought of as a storing bytes into the register, can it not?

Reply:
Let's maybe ignore the register issue for now :) Is the "FIXME" I suggested OK, or do we need to reconsider it?

Reply:
Having reviewed this thread, I think neither my original comment nor the proposed FIXME above is right.

SIR is abstract in the sense that it makes no assumptions about register allocation decisions. The code in question asks to access 8 bytes at an 8 byte offset of an abstract 16-byte value. It just so happens that for now that (for simplicity) the register allocator will put 16-byte values on the stack. But there's no reason a 16-byte value couldn't span two (arbitrary) registers at runtime. Rust already does things like this, probably for performance reasons.

Therefore a better way forward is to remove the comment altogether, as it's confusing.

Would you be happy with that?

Reply:
To be clear: if the register allocator completely changes how it lays out the 16 byte value, nothing in this piece of code will need to change?

Reply:
That's what I'm aiming at, yeah.

(It should also be possible to offset constants in the same way. Something else on my to-do list)

Reply:
Aiming at? Can we be a bit more precise please for the sake of my constitution :)

Reply:
Yes, that is how I would like to implement this. 😄

Reply:
Just to be clear (because I'm still not clear): nothing in this code will need to change, no matter what we change in the register allocator? Is that the case now, or the case that we would like it to be at some point in the future?

Reply:
Nothing in this code will need to change.

Reply:
OK, got it.

Reply:
Fixed in 2fb17a4.